### DISCUSSION TITLE
Resolving DNS using IPv4 DNS server not working despite NAT64 working correctly

### DISCUSSION THREAD

**Discussion by agners** [openthread/openthread#9782] [Q&A] [answered]
Hello,

We are trying to use the DNS resolver capabilities of the OTBR. When hosting the OTBR in an environment where the host has IPv6, resolving a DNS name works just fine:

```
root@ha-fujitsu:/usr/src# ot-ctl dns config
Server: [2001:4860:4860:0:0:0:0:8888]:53
ResponseTimeout: 6000 ms
MaxTxAttempts: 3
RecursionDesired: yes
ServiceMode: srv_txt_opt
Nat64Mode: allow
Done
root@ha-fujitsu:/usr/src# ot-ctl dns resolve4 example.com
DNS response for example.com. - fdf8:76ab:ebd6:2:0:0:5db8:d822 TTL:14385 
Done
```

However, trying to resolve using an IPv4 address doesn't seem to work:
```
root@ha-fujitsu:/usr/src# ot-ctl dns resolve4 example.com 8.8.8.8
Synthesized IPv6 DNS server address: fdf8:76ab:ebd6:2:0:0:808:808
DNS response for example.com. - 
Error 28: ResponseTimeout
```

Pinging the same IPv4 address seems to work however:
```
ot-ctl ping 8.8.8.8
Pinging synthesized IPv6 address: fdf8:76ab:ebd6:2:0:0:808:808
16 bytes from fdf8:76ab:ebd6:2:0:0:808:808: icmp_seq=1 hlim=111 time=3ms
1 packets transmitted, 1 packets received. Packet loss = 0.0%. Round-trip min/avg/max = 3/3.0/3 ms.
Done
```

I am quite certain that I don't have any firewall active which would prevent the IPv4 DNS request. So far I haven't further looked into the issue (I didn't analyze network traffic yet), but before digging deeper, I was wondering if this should be working in general.

I am also a bit confused about the DNS64 option. It seems that this simply adds a local bind server which forwards requests to IPv4 addresses. Is this really necessary? With NAT64, and from the description of the dns commands it sounds like OTBR can use IPv4 addresses, which will be synthesized to an IPv6 address using the preferred NAT64 prefix, just fine. So is this option required to reach IPv4 DNS server when NAT64 is already working fine?

**Comment by jwhui** [comment]
The `dns resolve4` command should work. If you can help grab pcaps for both the Thread and infrastructure interfaces, we can help take a look.

The `DNS64` option is not necessary any more. In Thread, the client is responsible for directly retrieving A records and synthesizing IPv6 addresses.

**Comment by agners** [comment, marked as answer]
Finally got to the bottom of this (two years later 😅). Root cause and fix: openthread/openthread#13553.

**Why `dns resolve4 example.com 8.8.8.8` times out while `ping` to the synthesized address works:**

The two commands take different paths. `ping` runs entirely inside the OpenThread stack — network data routing matches the NAT64 route and hands the packet straight to the translator. The DNS client, however, uses a **host kernel UDP socket** on POSIX border routers (`OPENTHREAD_CONFIG_PLATFORM_UDP_ENABLE`, with `DNS_CLIENT_BIND_UDP_TO_THREAD_NETIF=0` by default), so the **host routing table** decides where the query goes.

And the host has no route for the NAT64 prefix: the posix platform skips installing self-published external routes, and only manages the IPv4 CIDR route (192.168.255.0/24 → wpan0), never the IPv6-side /96. So the query to the NAT64-synthesized server address leaks out the IPv6 default route toward the LAN gateway and is silently lost — or fails with `sendmsg: Network is unreachable` on hosts without an IPv6 default route.

**Proof** (on a Home Assistant Yellow BR):

```
# ot-ctl dns resolve4 example.com 8.8.8.8
Synthesized IPv6 DNS server address: fd3d:86c4:8b3c:2:0:0:808:808
Error 28: ResponseTimeout

# ip -6 route add fd3d:86c4:8b3c:2::/96 dev wpan0    # prefix from `ot-ctl br nat64prefix`
# ot-ctl dns resolve4 example.com 8.8.8.8
DNS response for example.com. - fd3d:86c4:8b3c:2:0:0:6814:179a TTL:300 fd3d:86c4:8b3c:2:0:0:ac42:93f3 TTL:300
Done
```

The proposed fix installs this route automatically while the translator is active, mirroring the existing IPv4 CIDR route handling. Nice side effect: any host application on the BR can then use NAT64-synthesized addresses too.

(For completeness: `dns resolve4` on the BR against the *default* DNS server — the BR's own DNS-SD server — fails for a separate reason: that server is an OpenThread-internal socket, so there is no listener on kernel port 53 and the BR host cannot reach its own DNS-SD server. Queries from actual Thread devices are unaffected by either issue.)
